### PR TITLE
HDDS-9724. Provide stack trace for RocksObject allocation

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Managed RocksDB objects have optional leak detection (enabled in integration tests).  Objects that extend `ManagedObject` report leak with stack trace, but most other types lack this detail, making it more difficult to find the source of the leak:

```
[Finalizer] WARN  managed.ManagedRocksObjectUtils (ManagedRocksObjectUtils.java:reportLeak(71)) - ManagedWriteBatch is not closed properly
```

This change improves it by adding stack trace information for these objects.

https://issues.apache.org/jira/browse/HDDS-9724

## How was this patch tested?

Applied the patch on top of https://github.com/devmadhuu/ozone/commit/81fb066918893cea15e4f49bb1af71d7d971cde6, which was not closing `RDBBatchOperation`.  Ran `TestReconTasks`, checked log for stack trace:

```
[Finalizer] WARN  managed.ManagedRocksObjectUtils (ManagedRocksObjectUtils.java:reportLeak(67)) - ManagedWriteBatch is not closed properly
StackTrace for unclosed instance: org.apache.hadoop.hdds.utils.db.managed.ManagedWriteBatch.<init>(ManagedWriteBatch.java:32)
org.apache.hadoop.hdds.utils.db.RDBBatchOperation.<init>(RDBBatchOperation.java:332)
org.apache.hadoop.ozone.recon.TestReconTasks.testMissingContainerDownNode(TestReconTasks.java:157)
...
```

Regular CI:
https://github.com/adoroszlai/ozone/actions/runs/6929643804

### CODE REVIEW

File: hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedBloomFilter.java [hunk 1]
```diff
@@ -20,14 +20,20 @@
...
+import static org.apache.hadoop.hdds.utils.db.managed.ManagedRocksObjectUtils.assertClosed;
+import static org.apache.hadoop.hdds.utils.db.managed.ManagedRocksObjectUtils.formatStackTrace;
+import static org.apache.hadoop.hdds.utils.db.managed.ManagedRocksObjectUtils.getStackTrace;
+
 /**
  * Managed BloomFilter.
  */
 public class ManagedBloomFilter extends BloomFilter {
 
+  private final StackTraceElement[] elements = getStackTrace();
```
Comment (duongkame) [inline hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedBloomFilter.java:32]: I guess it's more obvious if stacktrace is an `Optional<StackTraceElement[]>` to it's not always available (disabled in prod).

**Review by duongkame — COMMENTED** [review]
Thanks for the patch @adoroszlai . LGTM overall, I just have a suggestion to make the optional stacktrace more obvious.

File: hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedBloomFilter.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedBloomFilter.java:32]: Good point, but let me annotate as `@Nullable` instead.

**Review by hemantk-12 — APPROVED** [review]
Thanks for the patch @adoroszlai 
LGTM.

**Comment by adoroszlai** [comment]
Thanks @duongkame, @hemantk-12 for the review.

**Pull request merged**